### ISSUE TITLE
Troubleshooting 

### ISSUE THREAD

**Issue by iris-ywang** [iris-ywang/chemvae_model#1]

**Comment by iris-ywang** [comment]
SA-VAE:
`x_in` = `x_i` -> [encoder] -> `z_i` -> [decoder] -> `x_out` =`x_i`

PA-VAE:
`x_in` = (`x_i`, `x_j`) -> [encoder] -> `z_i` -> [decoder] -> (`x_j`, `x_i`) -> [swap_layer] -> `x_out` = (`x_i`, `x_j`)

We have a 2-target-value outputs model. One for `x_out`, and one for `z_mean_log_loss`.

So for model prediction target, target `y` = `{x_out: x_in, z_mean_log_loss = 0.(of some shapes)}`

**Comment by iris-ywang** [comment]
Checked training structure. Didn't spot any issue. Need to check the model parameters - why is it always giving `nan`?

**Comment by iris-ywang** [comment]
zinc_paired_encoder.h5 was void. All params were NaN.

In zinc_paired_encoder2.h5, adding a `if` check for NaN in `X_train` and `X_test`.

**Comment by iris-ywang** [comment]
Tested `zinc_paired_encoder2.h5` with `epoch=5` for 5 batches. There is now a quick test in `chemvae.test_model.quick_test_encoder`. 

The results showed fine. There was no NaN weights in `zinc_paired_encoder2.h5`.

But in `zinc_paired_encoder.h5`, yes weights are NaN. So one of the batches messed up the whole model.

So for run 2:
- Remedy 1: save batch-wise model in `zinc_paired_encoder2_252.h5`, `zinc_paired_encoder2_604.h5`, etc. If anything goes wrong you can track it down.
- Remedy 2: print encoder validation messages throughout the experiments.

To do: 
- Remove all the encoders and decoders(`zinc_paired_encoder.h5` and ` zinc_paired_decoder.h5` are void. `zinc_paired_encoder2.h5` etc are test version so invalid.)

**Comment by iris-ywang** [comment]
In Run2, batch runs 1-5 were fine (`zinc_paired_encoder2_1512.h5` was all good without NaN values). After that, the model starts to contain NaN weights. My check on `X_train` and `X_test` didn't find they contain NaN inputs. 

To-do:
1. Find out what is the cause
2. Check how good is `zinc_paired_encoder2_1512.h5`

**Comment by iris-ywang** [comment]
Asked chatgpt for potential reasons. 

1. **Exploding Gradients**: If the gradients during backpropagation become too large, they can cause the weights to update to NaN values. This is common in deep networks or networks with long sequences.
- **Solution**: Implement gradient clipping to cap the gradients at a maximum value.
`optimizer = tf.keras.optimizers.Adam(learning_rate=1e-4, clipnorm=1.0)`

2. **Learning Rate**: A very high learning rate can cause the weights to change too drastically, leading to NaN values.
- **Solution**: Reduce the learning rate.
- `optimizer = tf.keras.optimizers.Adam(learning_rate=1e-4)`

3. **Numerical Instability**: Operations such as division by very small numbers, log of zero, or exponential of large numbers can lead to NaNs.
- **Solution**: Ensure numerical stability in all operations, possibly adding small values (epsilon) to denominators, using functions like tf.math.log1p instead of tf.math.log, etc.

4. **Initialization**: Poor weight initialization can sometimes lead to numerical issues during training.
- **Solution**: Use a robust initialization method like He initialization or Xavier initialization.
`initializer = tf.keras.initializers.HeNormal()  # or tf.keras.initializers.GlorotNormal()`
`model.add(tf.keras.layers.Dense(units=..., activation='relu', kernel_initializer=initializer))`

5. **Activation Functions**: Some activation functions like ReLU can cause issues if the learning rate is too high, leading to dead neurons (always zero) or NaNs.
- **Solution**: Consider using alternative activations or ensuring the learning rate is appropriately set.

6. **Data Issues** (**ignore for** now because we have one hot values): Even though you checked for NaNs in your input data, there might be other numerical issues like extremely large or small values that can destabilize training.
- **Solution**: Normalize or standardize your input data.

7. **Loss Function**: Certain loss functions might produce NaN values if the predicted values are not in an expected range (e.g., log loss with probabilities not in the [0, 1] range).
- **Solution**: Make sure the outputs are within the expected range for the loss function.
- Verify Activation Outputs and Loss:
```
# Ensure activation outputs are in expected range
outputs = model.predict(X_batch)
print("Max output:", np.max(outputs))
print("Min output:", np.min(outputs))
```

8. **Hardware and Precision**: Issues with floating-point precision can sometimes lead to NaNs, especially on certain hardware.
- **Solution**: Use higher precision (float64 instead of float32) if hardware allows, or check if specific hardware issues are known.

**Comment by iris-ywang** [comment]
**Attempt 1**: changed LR from 0.00039192162392520126 to 0.0001. 

Running Run2 from batch_id=6. 
Reloaded model is copied from `zinc_paired_encoder2_1512.h5`

**Comment by iris-ywang** [comment]
Attempt 1 didn't work. Further reducing to 5e-5

**Comment by iris-ywang** [comment]
LR=5e-5 still didn't work. Trying to still batch 6 and see if it's batch 6's problem

**Comment by iris-ywang** [comment]
No it was not batch 6's problem. 

Now **Attempt 2**: LR = 0.00039192162392520126 as before. Now `adam` as a clipnorm = 1.0 parameter.

**Comment by iris-ywang** [comment]
clipnorm = 1.0 is working!

**Comment by iris-ywang** [comment]
qsar for PA-VAE should be able to run. But the process got killed during the encoding process (not even got to the training). (PA QSAR running on 500 molecules only).

Potential reason: Memory kill

Todo:
1. batch the encoding process (rather than single molecule encoding)
2. batch the training with a small batch number.

**Comment by iris-ywang** [comment]
Completed the work to batch pairing&encoding and batch fitting.

**Comment by iris-ywang** [comment]
From Training batch: 46 (Training batch index is from 11592 to 11844, Test batch index is from 1150 to 1175), the model hits its limit again and starts to produce NaN weights.

Need to rerun from batch 46. **Attempt**: clipnorm = 0.5

**Comment by iris-ywang** [comment]
clipnorm = 0.5 failed.

**Attempt**: Further reducing it to clipnorm = 0.1

**Comment by iris-ywang** [comment]
OMG it worked! From Batch 46, clipnorm changed from 1.0 to 0.1.
